kernel/src/table_changes/mod.rs

```diff
@@ -0,0 +1,196 @@
+use crate::schema::{DataType, Schema, StructField, StructType};
+use crate::snapshot::Snapshot;
+use crate::table_features::ColumnMappingMode;
+use crate::{DeltaResult, Engine, Error, Version};
+
+static CDF_FIELDS: LazyLock<[StructField; 3]> = LazyLock::new(|| {
+    [
+        StructField::new("_change_type", DataType::STRING, false),
+        StructField::new("_commit_version", DataType::LONG, false),
+        StructField::new("_commit_timestamp", DataType::TIMESTAMP, false),
+    ]
+});
+
+/// Represents a call to read the Change Data Feed between two versions of a table. The schema of
+/// `TableChanges` will be the schema of the table with three additional columns:
+/// - `_change_type`: String representing the type of change that for that commit. This may be one
+///   of `delete`, `insert`, `update_preimage`, or `update_postimage`.
+/// - `_commit_version`: Long representing the commit the change occurred in.
+/// - `_commit_timestamp`: Time at which the commit occurred. If In-commit timestamps is enabled,
+///   this is retrieved from the [`CommitInfo`] action. Otherwise, the timestamp is the same as the
+///   commit file's modification timestamp.
+///
+/// For more details, see the following sections of the protocol:
+/// - [Add CDC File](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#add-cdc-file)
+/// - [Change Data Files](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#change-data-files).
+///
+/// [`CommitInfo`]: crate::actions::CommitInfo
+#[derive(Debug)]
+pub struct TableChanges {
+    pub log_segment: LogSegment,
+    table_root: Url,
+    end_snapshot: Snapshot,
+    start_version: Version,
+    schema: Schema,
+}
+
+impl TableChanges {
+    /// Creates a new [`TableChanges`] instance for the given version range. This function checks
+    /// these two properties:
+    /// - The change data feed table feature must be enabled in both the start or end versions.
+    /// - The schema at the start and end versions are the same.
+    ///
+    /// Note that this does not check that change data feed is enabled for every commit in the
+    /// range. It also does not check that the schema remains the same for the entire range.
+    ///
+    /// # Parameters
+    /// - `table_root`: url pointing at the table root (where `_delta_log` folder is located)
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `start_version`: The start version of the change data feed
+    /// - `end_version`: The end version of the change data feed. If this is none, this defaults to
+    ///   the newest table version.
+    pub fn try_new(
+        table_root: Url,
+        engine: &dyn Engine,
+        start_version: Version,
+        end_version: Option<Version>,
+    ) -> DeltaResult<Self> {
+        let start_snapshot =
+            Snapshot::try_new(table_root.as_url().clone(), engine, Some(start_version))?;
+        let end_snapshot = Snapshot::try_new(table_root.as_url().clone(), engine, end_version)?;
```

Comment:
I think we need to do `protocol.ensure_read_supported()`? and perhaps we should do that in `Snapshot::try_new()` and we could just have a comment here that says we rely on snapshot to check?

Reply:
related https://github.com/delta-io/delta-kernel-rs/pull/518